# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [201]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [202]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name,Transported
0,0001_01,Europa,False,B/0/P,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False
1,0002_01,Earth,False,F/0/S,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True
2,0003_01,Europa,False,A/0/S,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False
3,0003_02,Europa,False,A/0/S,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False
4,0004_01,Earth,False,F/1/S,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True


Now perform the same as before:
- Feature Scaling
- Feature Selection


In [203]:
#your code here

**Perform Train Test Split**

In [204]:
#your code here

**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

- Bagging and Pasting

In [205]:
#your code here

- Random Forests

In [206]:
#your code here

- Gradient Boosting

In [207]:
#your code here

- Adaptive Boosting

In [208]:
#your code here

Which model is the best and why?

In [209]:
#comment here

---
# Solved LAB | Ensemble Methods
---

### **Load & View the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [210]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [211]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name,Transported
0,0001_01,Europa,False,B/0/P,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False
1,0002_01,Earth,False,F/0/S,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True
2,0003_01,Europa,False,A/0/S,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False
3,0003_02,Europa,False,A/0/S,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False
4,0004_01,Earth,False,F/1/S,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True


In [212]:
spaceship.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


---
### **Data Cleaning**

#### **Check the shape of your data**

In [213]:
# Check the shape of the dataset*
print("Dataset Shape:", spaceship.shape)
# number of rows 
print("Number of rows:", spaceship.shape[0])
# number of columns
print("Number of columns:", spaceship.shape[1])

Dataset Shape: (8693, 14)
Number of rows: 8693
Number of columns: 14


#### **Check for data types**

In [214]:
print("Data Types:")
display(spaceship.dtypes)

Data Types:


PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

#### **Check for missing values**

In [215]:
print("Missing Values:")
display(spaceship.isnull().sum())

Missing Values:


PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continouos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value. 

#### **Deal with Nulls**

In [216]:
# handle missing values
'''
spaceship["HomePlanet"] = spaceship["HomePlanet"].fillna("Unknown") # fill missing values with "Unknown"
spaceship["CryoSleep"] = spaceship["CryoSleep"].fillna(False)
spaceship["Destination"] = spaceship["Destination"].fillna("Unknown") # fill missing values with "Unknown"
spaceship["Age"] = spaceship["Age"].fillna(spaceship["Age"].median()) # fill missing values with median before converting to int
spaceship["RoomService"] = spaceship["RoomService"].fillna(0) # fill missing values with 0
spaceship["FoodCourt"] = spaceship["FoodCourt"].fillna(0) # fill missing values with 0
spaceship["ShoppingMall"] = spaceship["ShoppingMall"].fillna(0) # fill missing values with 0
spaceship["Spa"] = spaceship["Spa"].fillna(0) # fill missing values with 0
spaceship["VRDeck"] = spaceship["VRDeck"].fillna(0) # fill missing values with 0
spaceship["Cabin"] = spaceship["Cabin"].fillna("Unknown") # fill missing values with "Unknown"
'''

# drop nulls in spaceship dataset
for col in spaceship.columns:
    print(f"Missing values in '{col}': {spaceship[col].isnull().sum()}")
    # if there are missing values, drop the rows with missing values in that column
    if spaceship[col].isnull().sum() > 0:
        spaceship = spaceship.dropna(subset=[col])
    print(f"Missing values in '{col}' after dropping nulls: {spaceship[col].isnull().sum()}")

Missing values in 'PassengerId': 0
Missing values in 'PassengerId' after dropping nulls: 0
Missing values in 'HomePlanet': 201
Missing values in 'HomePlanet' after dropping nulls: 0
Missing values in 'CryoSleep': 215
Missing values in 'CryoSleep' after dropping nulls: 0
Missing values in 'Cabin': 186
Missing values in 'Cabin' after dropping nulls: 0
Missing values in 'Destination': 171
Missing values in 'Destination' after dropping nulls: 0
Missing values in 'Age': 169
Missing values in 'Age' after dropping nulls: 0
Missing values in 'VIP': 179
Missing values in 'VIP' after dropping nulls: 0
Missing values in 'RoomService': 161
Missing values in 'RoomService' after dropping nulls: 0
Missing values in 'FoodCourt': 159
Missing values in 'FoodCourt' after dropping nulls: 0
Missing values in 'ShoppingMall': 173
Missing values in 'ShoppingMall' after dropping nulls: 0
Missing values in 'Spa': 161
Missing values in 'Spa' after dropping nulls: 0
Missing values in 'VRDeck': 154
Missing values 

#### **Deal with Duplicates**

In [217]:
# remove duplicates
spaceship = spaceship.drop_duplicates()

---
### **Data Preparation**
Now perform the same as before:
- Feature Scaling
- Feature Selection


#### Split **Cabin**  to "Deck", "CabinNumber", and "Side"

In [218]:
# check the unique values in the 'Cabin' column
print("unique values in 'Cabin' column before splitting:", spaceship['Cabin'].unique())

# split the 'Cabin' column into 'Deck', 'CabinNumber', and 'Side'
spaceship[['Deck', 'CabinNumber', 'Side']] = spaceship['Cabin'].str.split('/', expand=True)

# drop the original 'Cabin' column
spaceship = spaceship.drop('Cabin', axis=1)
print(100*'-')
# check the unique values in the new columns
print("unique values in 'Deck' column after splitting:", spaceship['Deck'].unique())
print("unique values in 'CabinNumber' column after splitting:", spaceship['CabinNumber'].unique())
print("unique values in 'Side' column after splitting:", spaceship['Side'].unique())

unique values in 'Cabin' column before splitting: ['B/0/P' 'F/0/S' 'A/0/S' ... 'G/1499/S' 'G/1500/S' 'E/608/S']


----------------------------------------------------------------------------------------------------
unique values in 'Deck' column after splitting: ['B' 'F' 'A' 'G' 'E' 'C' 'D' 'T']
unique values in 'CabinNumber' column after splitting: ['0' '1' '2' ... '1894' '1499' '608']
unique values in 'Side' column after splitting: ['P' 'S']


#### **Drop PassengerId and Name**

In [219]:
# drop PassengerId and Name columns since they are not useful for modeling
spaceship.drop(["PassengerId", "Name"], axis=1, inplace=True)

#### **Convert Data Types**

In [220]:
spaceship.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6606 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   HomePlanet    6606 non-null   object 
 1   CryoSleep     6606 non-null   object 
 2   Destination   6606 non-null   object 
 3   Age           6606 non-null   float64
 4   VIP           6606 non-null   object 
 5   RoomService   6606 non-null   float64
 6   FoodCourt     6606 non-null   float64
 7   ShoppingMall  6606 non-null   float64
 8   Spa           6606 non-null   float64
 9   VRDeck        6606 non-null   float64
 10  Transported   6606 non-null   bool   
 11  Deck          6606 non-null   object 
 12  CabinNumber   6606 non-null   object 
 13  Side          6606 non-null   object 
dtypes: bool(1), float64(6), object(7)
memory usage: 729.0+ KB


In [221]:
# convert Age to integer type
spaceship["Age"] = spaceship["Age"].astype(int)

# convert VIP to boolean type
spaceship["VIP"] = spaceship["VIP"].astype(bool)

# convert booleans to integers (True=1, False=0) for modeling
for col in spaceship.columns:
    if spaceship[col].dtype == 'bool':
        spaceship[col] = spaceship[col].astype(int)


---
### **Perform Train Test Split**
- split spaceship dataset into train and test sets (stratify by target variable to maintain class distribution)


**Define target (y)**

In [222]:
# define target variable
y = spaceship["Transported"]

**Define features (X)**

In [223]:
# define features: 
X = spaceship.drop(["Transported"], axis=1)
print(f"Shape of data after dropping 'Transported', 'PassengerId', 'Name': {X.shape}")
display(X.head())

Shape of data after dropping 'Transported', 'PassengerId', 'Name': (6606, 13)


,HomePlanet,CryoSleep,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Deck,CabinNumber,Side
0,Europa,False,TRAPPIST-1e,39,0,0.0,0.0,0.0,0.0,0.0,B,0,P
1,Earth,False,TRAPPIST-1e,24,0,109.0,9.0,25.0,549.0,44.0,F,0,S
2,Europa,False,TRAPPIST-1e,58,1,43.0,3576.0,0.0,6715.0,49.0,A,0,S
3,Europa,False,TRAPPIST-1e,33,0,0.0,1283.0,371.0,3329.0,193.0,A,0,S
4,Earth,False,TRAPPIST-1e,16,0,303.0,70.0,151.0,565.0,2.0,F,1,S


**SPLIT**

In [224]:
# define train-test split sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

#display the shapes of the resulting sets
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (5284, 13)
X_test shape: (1322, 13)
y_train shape: (5284,)
y_test shape: (1322,)


---
### **Feature Engineering**

1. Identity Categorical vs Numerical Features

In [225]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, StandardScaler

# Boolean columns (already 0/1)
bool_cols = ["CryoSleep", "VIP"]

# Categorical columns (more than 2 categories)
cat_cols = ["HomePlanet", "Destination", "Deck", "Side"]

# Numerical columns
num_cols = ["Age", "RoomService", "FoodCourt", "ShoppingMall", "Spa", "VRDeck", "CabinNumber"]

2. one hot encode categorical features (not boolean)

In [226]:
ohe = OneHotEncoder(handle_unknown="ignore", sparse_output=False)

X_train_cat = pd.DataFrame(
    ohe.fit_transform(X_train[cat_cols]),
    columns=ohe.get_feature_names_out(cat_cols),
    index=X_train.index
)

X_test_cat = pd.DataFrame(
    ohe.transform(X_test[cat_cols]),
    columns=ohe.get_feature_names_out(cat_cols),
    index=X_test.index
)

3. Scale Numerical Columns

In [227]:
scaler = StandardScaler()

X_train_num = pd.DataFrame(
    scaler.fit_transform(X_train[num_cols]),
    columns=num_cols,
    index=X_train.index
)

X_test_num = pd.DataFrame(
    scaler.transform(X_test[num_cols]),
    columns=num_cols,
    index=X_test.index
)

4. Convert Booleans to int (True=1 , False=0)

In [228]:
# Boolean columns (already 0/1)
X_train_bool = X_train[bool_cols]
X_test_bool = X_test[bool_cols]

5. Combine & Save Train Test sets CSVs

In [229]:
# Combine all features
X_train = pd.concat([X_train_num, X_train_bool, X_train_cat], axis=1)
X_test = pd.concat([X_test_num, X_test_bool, X_test_cat], axis=1)

# display the shapes of the resulting sets
print("X_train shape after encoding and scaling:", X_train.shape)
display(X_train.head())
print("X_test shape after encoding and scaling:", X_test.shape)
display(X_test.head())

print("y_train shape:", y_train.shape)
display(y_train.head())
print("y_test shape:", y_test.shape)
display(y_test.head())

# ---- Save preprocessed data to CSV files ----
X_train.to_csv("X_train.csv", index=False)
X_test.to_csv("X_test.csv", index=False)
y_train.to_csv("y_train.csv", index=False)
y_test.to_csv("y_test.csv", index=False)


X_train shape after encoding and scaling: (5284, 25)


,Age,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,CabinNumber,CryoSleep,VIP,HomePlanet_Earth,...,Deck_A,Deck_B,Deck_C,Deck_D,Deck_E,Deck_F,Deck_G,Deck_T,Side_P,Side_S
6017,0.280747,-0.342297,-0.263556,-0.299472,-0.274524,2.517793,-1.034191,False,0,0.0,...,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
8533,0.625019,1.909566,-0.286602,0.758795,-0.275405,-0.272281,0.004668,False,0,0.0,...,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0
6961,1.175855,-0.342297,-0.286602,-0.299472,-0.275405,-0.272281,-0.619814,True,0,0.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
2248,-0.476651,0.102591,-0.047047,-0.299472,-0.275405,-0.269638,-0.201547,False,0,1.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0
2422,-0.614360,-0.252405,-0.286602,2.461079,1.183715,-0.272281,-1.008900,False,0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0


X_test shape after encoding and scaling: (1322, 25)


,Age,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,CabinNumber,CryoSleep,VIP,HomePlanet_Earth,...,Deck_A,Deck_B,Deck_C,Deck_D,Deck_E,Deck_F,Deck_G,Deck_T,Side_P,Side_S
8153,0.280747,-0.342297,1.578894,-0.299472,-0.272763,0.603416,-0.956374,False,1,0.0,...,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
7374,-0.201234,-0.342297,-0.286602,-0.299472,-0.275405,-0.272281,1.317833,True,0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
8231,-1.716031,-0.342297,-0.286602,-0.299472,-0.275405,-0.272281,1.599921,True,0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0
5795,-1.784885,-0.342297,-0.286602,-0.299472,-0.275405,-0.272281,0.775059,True,0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0
2536,-0.338942,-0.342297,-0.286602,-0.299472,-0.275405,-0.272281,-0.966101,True,0,0.0,...,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0


y_train shape: (5284,)


6017    0
8533    0
6961    1
2248    0
2422    0
Name: Transported, dtype: int64

y_test shape: (1322,)


8153    0
7374    1
8231    0
5795    1
2536    1
Name: Transported, dtype: int64

---
### **Model Selection** 
now you will try to apply different ensemble methods in order to get a better model

---

**To improve model performance, several ensemble methods were applied and compared.**

---

#### 1️⃣ **Bagging and Pasting**

- Bagging (Bootstrap Aggregating) trains multiple base models on random subsets of the data.
- Bagging: Sampling with replacement
- Pasting: Sampling without replacement
- Base learner: Decision Tree

These methods reduce variance and help prevent overfitting.


#### **Bagging (with replacement)**

In [230]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    average_precision_score,
    roc_auc_score,
    confusion_matrix
)

# Base model
dt = DecisionTreeClassifier(random_state=42)

# Bagging
bagging = BaggingClassifier(
    estimator=dt,
    n_estimators=100,
    bootstrap=True,
    random_state=42,
    n_jobs=-1
)

bagging.fit(X_train, y_train)

# Predictions
y_train_pred = bagging.predict(X_train)
y_test_pred = bagging.predict(X_test)

y_train_proba = bagging.predict_proba(X_train)[:, 1]
y_test_proba = bagging.predict_proba(X_test)[:, 1]

# -----------------------
# Important Metrics
# -----------------------

print("=== BAGGING RESULTS ===")

# Accuracy
print("Train Accuracy:", accuracy_score(y_train, y_train_pred))
print("Test Accuracy :", accuracy_score(y_test, y_test_pred))

# Classification Report
print("\nClassification Report:\n")
print(classification_report(y_test, y_test_pred))

# PR AUC
print("PR AUC:", average_precision_score(y_test, y_test_proba))

# ROC AUC
print("Train ROC AUC:", roc_auc_score(y_train, y_train_proba))
print("Test ROC AUC :", roc_auc_score(y_test, y_test_proba))

# Confusion Matrix
print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_test_pred))

# Overfitting Check
print("\nAccuracy Gap:", 
      accuracy_score(y_train, y_train_pred) - accuracy_score(y_test, y_test_pred))

print("ROC AUC Gap:", 
      roc_auc_score(y_train, y_train_proba) - roc_auc_score(y_test, y_test_proba))

=== BAGGING RESULTS ===
Train Accuracy: 0.9998107494322483
Test Accuracy : 0.8018154311649016

Classification Report:

              precision    recall  f1-score   support

           0       0.78      0.83      0.81       656
           1       0.82      0.77      0.80       666

    accuracy                           0.80      1322
   macro avg       0.80      0.80      0.80      1322
weighted avg       0.80      0.80      0.80      1322

PR AUC: 0.9010410076421606
Train ROC AUC: 0.9999987822005865
Test ROC AUC : 0.8853628781220246

Confusion Matrix:

[[546 110]
 [152 514]]

Accuracy Gap: 0.19799531826734673
ROC AUC Gap: 0.11463590407856183


#### **Pasting (without replacement)**

In [231]:
pasting = BaggingClassifier(
    estimator=dt,
    n_estimators=100,
    bootstrap=False,
    random_state=42,
    n_jobs=-1
)

pasting.fit(X_train, y_train)

y_train_pred = pasting.predict(X_train)
y_test_pred = pasting.predict(X_test)

y_train_proba = pasting.predict_proba(X_train)[:, 1]
y_test_proba = pasting.predict_proba(X_test)[:, 1]

print("=== PASTING RESULTS ===")

print("Train Accuracy:", accuracy_score(y_train, y_train_pred))
print("Test Accuracy :", accuracy_score(y_test, y_test_pred))

print("\nClassification Report:\n")
print(classification_report(y_test, y_test_pred))

print("PR AUC:", average_precision_score(y_test, y_test_proba))

print("Train ROC AUC:", roc_auc_score(y_train, y_train_proba))
print("Test ROC AUC :", roc_auc_score(y_test, y_test_proba))

print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_test_pred))

print("\nAccuracy Gap:", 
      accuracy_score(y_train, y_train_pred) - accuracy_score(y_test, y_test_pred))

print("ROC AUC Gap:", 
      roc_auc_score(y_train, y_train_proba) - roc_auc_score(y_test, y_test_proba))

=== PASTING RESULTS ===
Train Accuracy: 0.9998107494322483
Test Accuracy : 0.7549167927382754

Classification Report:

              precision    recall  f1-score   support

           0       0.76      0.73      0.75       656
           1       0.75      0.78      0.76       666

    accuracy                           0.75      1322
   macro avg       0.76      0.75      0.75      1322
weighted avg       0.76      0.75      0.75      1322

PR AUC: 0.738637372148855
Train ROC AUC: 0.9999999283647404
Test ROC AUC : 0.789088249835201

Confusion Matrix:

[[481 175]
 [149 517]]

Accuracy Gap: 0.24489395669397296
ROC AUC Gap: 0.2109116785295394


#### **2️⃣ Random Forests**

Random Forest is an extension of Bagging that:
- Uses Decision Trees
- Applies bootstrap sampling
- Randomly selects features at each split

This improves generalization and reduces correlation between trees.

In [232]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

rf.fit(X_train, y_train)

y_train_pred = rf.predict(X_train)
y_test_pred = rf.predict(X_test)

y_train_proba = rf.predict_proba(X_train)[:, 1]
y_test_proba = rf.predict_proba(X_test)[:, 1]

print("=== RANDOM FOREST RESULTS ===")

print("Train Accuracy:", accuracy_score(y_train, y_train_pred))
print("Test Accuracy :", accuracy_score(y_test, y_test_pred))

print("\nClassification Report:\n")
print(classification_report(y_test, y_test_pred))

print("PR AUC:", average_precision_score(y_test, y_test_proba))

print("Train ROC AUC:", roc_auc_score(y_train, y_train_proba))
print("Test ROC AUC :", roc_auc_score(y_test, y_test_proba))

print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_test_pred))

print("\nAccuracy Gap:",
      accuracy_score(y_train, y_train_pred) - accuracy_score(y_test, y_test_pred))

print("ROC AUC Gap:",
      roc_auc_score(y_train, y_train_proba) - roc_auc_score(y_test, y_test_proba))

=== RANDOM FOREST RESULTS ===
Train Accuracy: 0.9998107494322483
Test Accuracy : 0.7987897125567323

Classification Report:

              precision    recall  f1-score   support

           0       0.78      0.83      0.80       656
           1       0.82      0.76      0.79       666

    accuracy                           0.80      1322
   macro avg       0.80      0.80      0.80      1322
weighted avg       0.80      0.80      0.80      1322

PR AUC: 0.9003160032373015
Train ROC AUC: 0.999999641823702
Test ROC AUC : 0.8868449241924852

Confusion Matrix:

[[547 109]
 [157 509]]

Accuracy Gap: 0.2010210368755161
ROC AUC Gap: 0.11315471763121687


#### **3️⃣ Gradient Boosting**

Gradient Boosting builds trees sequentially:
- Each new tree corrects the errors of the previous one
- Focuses on minimizing a loss function
- Reduces bias

It typically achieves higher predictive accuracy.

In [233]:
from sklearn.ensemble import GradientBoostingClassifier

gb = GradientBoostingClassifier(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

gb.fit(X_train, y_train)

y_train_pred = gb.predict(X_train)
y_test_pred = gb.predict(X_test)

y_train_proba = gb.predict_proba(X_train)[:, 1]
y_test_proba = gb.predict_proba(X_test)[:, 1]

print("=== GRADIENT BOOSTING RESULTS ===")

print("Train Accuracy:", accuracy_score(y_train, y_train_pred))
print("Test Accuracy :", accuracy_score(y_test, y_test_pred))

print("\nClassification Report:\n")
print(classification_report(y_test, y_test_pred))

print("PR AUC:", average_precision_score(y_test, y_test_proba))

print("Train ROC AUC:", roc_auc_score(y_train, y_train_proba))
print("Test ROC AUC :", roc_auc_score(y_test, y_test_proba))

print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_test_pred))

print("\nAccuracy Gap:",
      accuracy_score(y_train, y_train_pred) - accuracy_score(y_test, y_test_pred))

print("ROC AUC Gap:",
      roc_auc_score(y_train, y_train_proba) - roc_auc_score(y_test, y_test_proba))

=== GRADIENT BOOSTING RESULTS ===
Train Accuracy: 0.8296744890234671
Test Accuracy : 0.8040847201210287

Classification Report:

              precision    recall  f1-score   support

           0       0.83      0.76      0.79       656
           1       0.78      0.85      0.81       666

    accuracy                           0.80      1322
   macro avg       0.81      0.80      0.80      1322
weighted avg       0.81      0.80      0.80      1322

PR AUC: 0.9120267229141965
Train ROC AUC: 0.9256837907889377
Test ROC AUC : 0.9016413517175712

Confusion Matrix:

[[498 158]
 [101 565]]

Accuracy Gap: 0.02558976890243836
ROC AUC Gap: 0.024042439071366495


#### **4️⃣ Adaptive Boosting (AdaBoost)**

AdaBoost:
- Trains weak learners sequentially
- Increases weight on misclassified samples
- Combines models using weighted voting

It improves performance by focusing on difficult observations.

In [234]:
from sklearn.ensemble import AdaBoostClassifier

ada = AdaBoostClassifier(
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

ada.fit(X_train, y_train)

y_train_pred = ada.predict(X_train)
y_test_pred = ada.predict(X_test)

y_train_proba = ada.predict_proba(X_train)[:, 1]
y_test_proba = ada.predict_proba(X_test)[:, 1]

print("=== ADABOOST RESULTS ===")

print("Train Accuracy:", accuracy_score(y_train, y_train_pred))
print("Test Accuracy :", accuracy_score(y_test, y_test_pred))

print("\nClassification Report:\n")
print(classification_report(y_test, y_test_pred))

print("PR AUC:", average_precision_score(y_test, y_test_proba))

print("Train ROC AUC:", roc_auc_score(y_train, y_train_proba))
print("Test ROC AUC :", roc_auc_score(y_test, y_test_proba))

print("\nConfusion Matrix:\n")
print(confusion_matrix(y_test, y_test_pred))

print("\nAccuracy Gap:",
      accuracy_score(y_train, y_train_pred) - accuracy_score(y_test, y_test_pred))

print("ROC AUC Gap:",
      roc_auc_score(y_train, y_train_proba) - roc_auc_score(y_test, y_test_proba))

=== ADABOOST RESULTS ===
Train Accuracy: 0.7469719909159728
Test Accuracy : 0.7299546142208775

Classification Report:

              precision    recall  f1-score   support

           0       0.68      0.86      0.76       656
           1       0.81      0.60      0.69       666

    accuracy                           0.73      1322
   macro avg       0.75      0.73      0.73      1322
weighted avg       0.75      0.73      0.73      1322

PR AUC: 0.8249109623222503
Train ROC AUC: 0.8513073363245353
Test ROC AUC : 0.834081795209844

Confusion Matrix:

[[563  93]
 [264 402]]

Accuracy Gap: 0.017017376695095288
ROC AUC Gap: 0.017225541114691212


---
### **Compare Results**
Which model is the best and why?

**🎯 Metrics I Prioritized for Classification, I prioritize:**

1️⃣ Test ROC AUC (Most Important)
- Measures ranking quality
- Threshold independent
- Best overall indicator of model quality

2️⃣ Test Accuracy
- Easy interpretability
- Good when classes are balanced (yours are balanced)

3️⃣ Overfitting Gap (Train vs Test)
- Small gap = better generalization
- Large gap = overfitting
- PR AUC and F1 were supportive but secondary.


---
- Bagging & Random Forest almost perfectly fit training data (Train ROC ≈ 1.0)
- That’s a classic sign of deep trees overfitting
- Gradient Boosting avoided that and generalized bette

- Gradient Boosting performed best based on highest Test ROC AUC, PR AUC, and smallest overfitting gap.
- Random Forest and Bagging performed similarly but showed strong overfitting.
- Pasting performed worst across all evaluation metrics.
---

| Model               | Test Acc | Test ROC | PR AUC | Overfitting | Rank |
|---------------------|----------|----------|--------|-------------|------|
| Gradient Boosting   | *0.804*    | *0.902*    | *0.912*  | Very Low    | 🥇   |
| Random Forest       | 0.799    | 0.887    | 0.900  | High        | 🥈   |
| Bagging             | 0.802    | 0.885    | 0.901  | High        | 🥉   |
| Pasting             | 0.755    | 0.789    | 0.739  | Very High   | 4️⃣   |

---
#### **🏆 Final Ranking (Best → Worst)**
---
**🥇 Gradient Boosting → BEST**
- Highest Test ROC AUC: 0.902
- Highest PR AUC: 0.912
- Highest Test Accuracy: 0.804
- Very small overfitting gap (~0.026)
- Balanced precision/recall
- This model generalizes extremely well.

Conclusion: Most stable + best predictive power.

---
**🥈 Random Forest**
- Test ROC AUC: 0.887
- Test Accuracy: 0.799
- PR AUC: 0.900
- Large overfitting (ROC gap ~0.113)
- Very strong performance but clearly overfitting.
---

**🥉 Bagging**
- Test ROC AUC: 0.885
- Test Accuracy: 0.802
- PR AUC: 0.901
- Large overfitting (ROC gap ~0.115)
- Very similar to Random Forest but slightly worse ROC.
---
**4️⃣ Pasting → WORST**
- Test ROC AUC: 0.789
- Test Accuracy: 0.755
- PR AUC: 0.739
- Massive overfitting (ROC gap ~0.211)
- Weakest model.